File: solutions/typescript/035-permutations.ts
/**
 * Problem 35: Permutations (LeetCode 46)
 * Difficulty: Med
 * Language: TypeScript
 */
function permute(nums: number[]): number[][] {
  if (nums.length === 0) return [[]];
  return nums.flatMap((n, i) => {
    const rest = [...nums.slice(0, i), ...nums.slice(i + 1)];
    return permute(rest).map(p => [n, ...p]);
  });
}

File: solutions/elixir/035-permutations.ex
# Problem 35: Permutations (LeetCode 46)
# Difficulty: Med
# Language: Elixir
# 
defmodule Permutations do
  def permute([]), do: [[]]
  def permute(nums) do
    Enum.flat_map(Enum.with_index(nums), fn {n, i} ->
      rest = List.delete_at(nums, i)
      Enum.map(permute(rest), fn p -> [n | p] end)
    end)
  end
end
